src/tomato/drivers/biologic/main.py

```diff
@@ -2,16 +2,94 @@
 import multiprocessing
 import time
 from filelock import FileLock
-
 from datetime import datetime, timezone
-
 from .kbio_wrapper import (
     get_kbio_techpath,
     payload_to_ecc,
     parse_raw_data,
     get_kbio_api,
 )
 
+def safe_api_connect(api,
+                    address: str,
+                    retries: int = 100,
+                    timeout: int = 10
+                    ) -> tuple:
+    """
+    Attempt to establish a connection with the device, retrying if necessary.
+
+    This function attempts to connect to the device at the specified address.
+    If the connection attempt fails, it retries up to a specified number of times,
+    waiting for a specified timeout period between each attempt.
+
+    Parameters
+    ----------
+    api
+        The API object used for making the connection.
+    address : str
+        The IP address of the device to connect to.
+    retries : int, optional
+        The number of times to retry the connection attempt. Default is 100.
+    timeout : int, optional
+        The time in seconds to wait between retries. Default is 10.
+
+    Returns
+    -------
+    tuple
+        A tuple containing the ID and device information upon successful connection.
+
+    Raises
+    ------
+    Exception
+        If the function fails to connect after the specified number of retries.
+
+    """
+    for _ in range(retries):
+        try:
+            id_, device_info = api.Connect(address)
+            return id_, device_info
+        except Exception as e:
+            time.sleep(timeout)
```

Comment:
The only thing I worry about with `sleep` here, is that I think we keep the device lock on, which might lead to time-outs there. Please check the interplay. It might be prudent to release the lock and then reacquire it after the sleep period.

Reply:
Do you think is it safer if I remove the sleep here? In this case, if the connection issue is just a glitch in the connection (whatever causes it). Then, multiple retires should be sufficient to solve this. This will ensure that we don't create new issue with device locking here. Say I try without sleep here. Then, if this doesn't solve the issue, then I will re-introduce the sleep function here again and will implement the lock-releasing and reacquire the lock as you suggested.

Reply:
I am pretty certain that when you sleep this function for more than 60 seconds, every other job trying to get a lock will fail, see: 

https://github.com/dgbowl/tomato/blob/56e9923d4f532e235054085e8c9ec43506e73061/src/tomato/drivers/biologic/main.py#L48

How you deal with this is up to you; you could also wrap the whole `with Filelock ...` block in a try-except-retries cycle, instead of just the connection. Might want to catch the particular Exception type raised, though, instead of a blank `except Exception as e:` then.

Reply:
I made a modification by wrapping FileLock into safe_api_connect and safe_api_disconnect function as suggested.